File: i3/.config/i3/unsplash-wallpaper.rb
#!/usr/bin/env ruby

require 'digest'
require 'open-uri'
require 'tmpdir'

SEED = Digest::SHA256.hexdigest('unsplash-wallpaper-v1').freeze

def save_to_tempfile(url)
  fullpath = File.join(Dir.tmpdir, SEED)
  return File.open(fullpath) if File.file?(fullpath)

  URI.open(url) do |image|
    content = image.read
    File.open(fullpath, 'wb+') do |f|
      f.binmode
      f.write(content)
      f
    end
  end
end

f = save_to_tempfile "https://source.unsplash.com/featured/3840x2160/daily/?cat"
`feh --bg-scale #{f.path}`

File: i3/.config/i3/unsplash-wallpaper.py
#!/usr/bin/env python

import os
import hashlib

from pathlib import Path
from urllib import request

def save_to_tempfile(url: str):
    p = Path(f'/tmp/{hashlib.sha256(url.encode()).hexdigest()}')
    abs_path = p.absolute()
    if not p.exists():
        with open(abs_path, 'wb') as f:
            f.write(request.urlopen(url).read())
    return abs_path

file_name = save_to_tempfile("https://source.unsplash.com/featured/3840x2160/daily/?cat")
os.system(f'/usr/bin/env feh --bg-scale {file_name}')
